### ISSUE TITLE
Scripts are flaky after ajax calls, waitForNavigation

### ISSUE THREAD

**Issue by sguptatw** [getgauge/taiko#168]
**Expected behavior**
Should be able to execute test scripts seamlessly

**Actual behavior**
Actions after waits like an ajax call or waitForNavigation is flaky

**Steps to replicate**
> Register on the site and use your credentials for the following script
* Run the script
```
(async () => {
    try {
        await openBrowser({headless:false});
        await goto("thetrainline.com",{timeout:20000});
 
        await write("London",into(inputField(toRightOf("From"))));
        await press("Enter");
        await write("Manchester");
        await press("Enter");
        await click("Tomorrow")
        await click(button({"type":"submit"}));
        await click("Continue");

        await click(checkBox(toLeftOf("Table Seat")));
        await click("Continue");
        await write("yourUserName@gmail.com",into(inputField({"placeHolder":"Email address"})));
        await write("yourPassword",into(inputField({"placeHolder":"Password"})));

        await click("Continue");
        await screenshot();
 
        await click(contains("Continue"));
        await click(contains("Proceed with")); 
    } catch (e) {
        console.error(e);
    } finally {
        closeBrowser();
    }
 })();
```
> Run this a couple of times in the non-observe mode. The script is inconsistent.

**Version**
```
commit - a9ea9fa69d7676775fc94be0e4c176b746fb7be5
```
Blocks #73
Blocks https://github.com/getgauge/taiko/issues/163

**Comment by NivedhaSenthil** [comment]
For me the script consistently fails giving, 
`Error: Element with text Table Seat not found` after clicking `continue` and the failure seems to be for valid reason. Not sure how it is related to #73, could see elements getting highlighted.

**Comment by sguptatw** [comment]
@NivedhaSenthil For some tickets the reservation element comes up that is when the above script fails. Here I have modified the script to take care of the flow change.

This was working earlier. Now when I run it multiple times this script with the same flow the test is flaky.

> Kindly note you will have to use your registered email and password in the script.
```
(async () => {
    try {
        await openBrowser({headless:false});
        await goto("thetrainline.com",{timeout:20000});
 
        await write("London",into(inputField(toRightOf("From"))));
        await press("Enter");
        await write("Manchester");
        await press("Enter");
        await click("Tomorrow")
        await click(button({"type":"submit"}));
        await click("Continue");

        if(await text("Yes").exists()==='Exists')
            click(await radioButton(toLeftOf("Yes")))

        await click(checkBox(toLeftOf("Table Seat")));
        await click("Continue");
        await write("yourRegisteredEmail@gmail.com",into(inputField({"placeHolder":"Email address"})));
        await write("yourPassword",into(inputField({"placeHolder":"Password"})));

        await click("Continue");
        await screenshot();
 
        await click(contains("Continue"));
        await click(contains("Proceed with")); 
    } catch (e) {
        console.error(e);
    } finally {
        closeBrowser();
    }
 })();
```

**Comment by sguptatw** [comment]
An example, where the script fails after an ajax request
```
(async () => {
   try {
       await openBrowser({headless:false});
       await goto("thetrainline.com",{timeout:20000});

       await write("London",into(inputField(toRightOf("From"))));
       await press("Enter");
       await write("Manchester");
       await press("Enter");
       await click("Tomorrow")
       await click(button({"type":"submit"}));
       await click(radioButton(toLeftOf(contains("$"))))
       await click("Continue");

       if(await text("Yes").exists()==='Exists')
           click(await radioButton(toLeftOf("Yes")))

       await click(checkBox(toLeftOf("Table Seat")));
       await click("Continue");
       await write("yourRegisteredEmail@gmail.com",into(inputField({"placeHolder":"Email address"})));
       await write("yourPassword",into(inputField({"placeHolder":"Password"})));

       await click("Continue");
       await screenshot();

       await click(contains("Continue"));
       await click(contains("Proceed with"));
   } catch (e) {
       console.error(e);
   } finally {
       closeBrowser();
   }
})();
```
> The output looks like
```
✔ Browser opened
 ✔ Navigated to url "http://thetrainline.com"
 ✔ Wrote London into the input Field To Right of From
 ✔ Pressed the Enter key
 ✔ Wrote Manchester into the focused element.
 ✔ Pressed the Enter key
 ✔ Clicked element containing text "Tomorrow"
 ✔ Clicked button[type="submit"]
 ✔ Clicked radio Button To Left of Element containing text "$"
Error: Element containing text "Continue" is coverred by other element
    at click (.../taiko/lib/taiko.js:364:15)
```

**Comment by NivedhaSenthil** [comment]
Both the cases are happening because the `xhr` and `frameLoad` events are happening after `500ms`, which is default time we wait to see if there is any `xhr` or `frameLoad` happening, only if it happens within that time we wait for it to complete. In both cases there is a `fetch` request happening immediately after click we might have to listen for any request in this case not just `xhr`.

One way to tell taiko to wait for `xhr` event to start for more than 500ms is to set `waitForStart` option,
```await click(radioButton(toLeftOf(contains("$"))),{waitForStart:2000})```
```await click(contains("Continue"),{waitForStart:2000});```

**Comment by sguptatw** [comment]
`500ms` seems to be too small for timing out. IMO - The default timeout for xhr should be increased say about to `15 secs`. Anyways if the calls complete before that the next command execute without waiting for timeout to expire. So increasing the default timeout for xhr will help.

**Comment by sriv** [comment]
15 secs is too much IMO. 2 second load time is a rule I've heard for SEO reasons. 

Perhaps leave it as a configuration per project?

**Comment by sguptatw** [comment]
On second thoughts `15 secs` may be too much. Assuming `2 second load` time, was thinking if the default timeout value can be about `5 secs` instead of `500ms` which seems too less. 

Taiko scripts can be run independently, not part of any project. The individual script step can have a timeout configuration. However, if we want to configure it overall, then a command line parameter may be?

**Comment by NivedhaSenthil** [comment]
500ms is not the timeout, it is the time we wait for any xhr or frameLoad to happen after any event like click...so that we can wait for that particular xhr or frameLoad to complete, timeout for those event to complete is my default 15s.  
Making that waitforStart to 5seconds by default will hit the performance of taiko as it will always wait for 5secs even if there is no xhr or frameLoad gonna happen. And in most cases xhr or frameLoads are triggered instantaneously, in this particular case there is other 'fetch' request which takes time to start xhr or frameLoads. 
- We might have to look for any request after an event and wait for it to complete instead of just `xhr`.   
- Will we good if `waitForStart` can be set globally, but it will also affect the performance of other events for which it may not be required.

**Comment by sguptatw** [comment]
**Update of the discussion**
To check if the element is covered by another element during exist. If it is covered, then 
- the exist method should wait until its no longer covered or timeout of the exist method.

**Comment by NivedhaSenthil** [comment]
waiting for element to be not covered cannot be done because if there are more than one element and the first element is covered we do click on the next element which is not covered… It will be tricky to find and wait for the right element in that case.

Improving wait to do below things covers most cases,
- Wait for networkIdle on navigation
- Wait for firstMeaningfulPaint on firstPaint
- Wait for all request triggered by user gesture

Fix done in 435a9ce.

Note: Open issues in chromium where `requestToBeSent` and `responseReceived` events not getting triggered which might affect in some cases as we relay on them to wait for requests. https://bugs.chromium.org/p/chromium/issues/detail?id=750469, https://bugs.chromium.org/p/chromium/issues/detail?id=883475&desc=2
